projects/extension/ai/semantic_catalog.py

```diff
@@ -0,0 +1,441 @@
+import json
+from collections.abc import Generator
+from typing import TypedDict
+
+
+class GeneratedDescription(TypedDict):
+    name: str
+    description: str
+
+
+def render_sample(plpy, relation: str, total: int = 5) -> str:
+    ident = ".".join([plpy.quote_ident(part) for part in relation.split(".")])
+    result = plpy.execute(f"select * from {ident}", total)
+
+    ret_obj = f"""<data id="{relation}">"""
+    for r in result:
+        values = []
+        for v in r.values():
+            if isinstance(v, str) or v is None:
+                values.append(plpy.quote_nullable(v))
+            elif isinstance(v, bool):
+                values.append("true" if v else "false")
+            else:
+                values.append(str(v))
+        ret_obj += f"""\n  insert into {ident} ({', '.join(plpy.quote_ident(key) for key in r.keys())}) values ({', '.join(values)});"""
+    ret_obj += "\n</data>"
+
+    return ret_obj
```

Comment:
I still think there are issues here, but let's merge and fix later. I think using SQL to render will be the most bulletproof approach.